src/lib.rs

```diff
@@ -827,23 +835,93 @@ impl SyncClient {
         }
     }
 
+    async fn request_wildcard_handler(
+        wrapped_client: Arc<Mutex<Self>>,
+        mut subscription_rx: MpscReceiver<Msg>,
+    ) {
+        let disconnecting = Self::disconnecting(Arc::clone(&wrapped_client));
+        pin_mut!(disconnecting);
+        loop {
+            // Select between the next message and disconnecting
+            let msg = match future::select(subscription_rx.next(), disconnecting).await {
+                Either::Left((Some(msg), unresolved_disconnecting)) => {
+                    disconnecting = unresolved_disconnecting;
+                    msg
+                }
+                Either::Left((None, _)) => break,
+                Either::Right(((), _)) => break,
+            };
+
+            let mut client = wrapped_client.lock().await;
+            let response_inbox = msg.subject().to_string();
+            if let Some(mut requester_tx) = client.request_inbox_mapping.remove(&response_inbox) {
+                requester_tx.send(msg).await.unwrap_or_else(|err| {
+                    warn!("Could not write response to pending request via mapping channel. Skipping! Err: {}", err);
+                });
+            }
+        }
+
+        // At this point we are either disconnecting or we're in some
+        // strange state where the subscription's rx has been closed.
+        let mut client = wrapped_client.lock().await;
+        client.request_inbox_mapping.clear();
+        client.request_wildcard_subscription = None;
+    }
+
     async fn request(
         wrapped_client: Arc<Mutex<Self>>,
         subject: &Subject,
         payload: &[u8],
     ) -> Result<Msg> {
         let inbox_uuid = Uuid::new_v4();
-        let reply_to = format!("{}.{}", util::INBOX_PREFIX, inbox_uuid.to_simple()).parse()?;
-        let mut rx = {
+
+        let (mut rx, reply_to) = {
             let mut client = wrapped_client.lock().await;
-            let (sid, rx) = client.subscribe(&reply_to, 1).await?;
-            client.unsubscribe_with_max_msgs(sid, 1).await?;
+            let request_inbox = inbox_uuid.to_simple();
+            let reply_to: Subject = format!(
+                "{}.{}.{}",
+                util::INBOX_PREFIX,
+                client.request_base_inbox,
+                request_inbox
+            )
+            .parse()?;
+
+            // Only subscribe to the wildcard subscription when requested once!
+            if client.request_wildcard_subscription.is_none() {
```

Comment:
We can probably move this to the connect method. I figured it might be a little bit better here so users won't find a random wildcard subscription each time they connect. TODO: Check to see what the nats people do in nats.go.

Reply:
I like it here. It looks like it is what [nats.go does](https://github.com/nats-io/nats.go/blob/6eb6122d3827c87e99b9a095ef3667d5aedf3b37/nats.go#L2647).